### ISSUE TITLE
RandomAuRa now supports working with more than one validator's node

### ISSUE THREAD

**Issue by varasev** [poanetwork/parity-ethereum#72]
Let's store the secret's cipher along with secret's hash in `RandomAuRa` contract: https://github.com/poanetwork/pos-contracts/blob/882b9bea93e76133115a82c2d39633dd45004013/contracts/RandomAuRa.sol#L24

On `commits phase`, when a node generates random `secret`, it hashes the `secret` with keccak256 and encrypts it with validator's private key.

Then the node calls the `commitHash()` and passes the hash and the cipher there. The cipher (along with the hash) is stored in `RandomAuRa` contract: https://github.com/poanetwork/pos-contracts/blob/882b9bea93e76133115a82c2d39633dd45004013/contracts/RandomAuRa.sol#L35

On `reveals phase` the node reads the cipher with `getCipher()` getter: https://github.com/poanetwork/pos-contracts/blob/882b9bea93e76133115a82c2d39633dd45004013/contracts/RandomAuRa.sol#L187

Then the node decrypts the cipher with validator's private key, gets the `secret` as a result, and calls `revealSecret()`.

What happens if the validator launches two nodes:

On `commits phase` the `node A` calls the `commitHash` and the `node B` calls the `commitHash` (they don't know about each other).

One of these calls will be approved, the other call will be silently ignored: https://github.com/poanetwork/pos-contracts/blob/882b9bea93e76133115a82c2d39633dd45004013/contracts/RandomAuRa.sol#L33

On `reveals phase` the `node A` gets the cipher, decrypts it, and calls `revealSecret`. The `node B` does the same, but the call to `revealSecret` is silently ignored because the `node A` has already revealed the secret: https://github.com/poanetwork/pos-contracts/blob/882b9bea93e76133115a82c2d39633dd45004013/contracts/RandomAuRa.sol#L51

So, if one of the nodes goes offline, the second node will be able to commit and/or reveal.

**Comment by afck** [comment]
That makes sense! :+1:
Since it's not obvious to me how to encrypt something in Parity from inside `AuthorityRound`, I'll start digging into the code and taking notes here.

Ethereum uses [ECDSA signatures with the secp256k1 curve](https://github.com/rust-bitcoin/rust-secp256k1), and [ethkey implements ECIES](https://github.com/poanetwork/parity-ethereum/blob/c159028507b8a35e1e39d6c8a11f4c72bee0a27e/ethkey/src/crypto.rs#L69), i.e. encryption and decryption using the same kind of key. This is used e.g. in `whisper`, `rpc` and `secret_store`, but nowhere in `ethcore` yet.

[`AuthorityRound`](https://github.com/poanetwork/parity-ethereum/blob/aura-pos/ethcore/src/engines/authority_round/mod.rs) has an [`EngineSigner`](https://github.com/poanetwork/parity-ethereum/blob/c159028507b8a35e1e39d6c8a11f4c72bee0a27e/ethcore/src/engines/signer.rs#L25), with methods to sign but not encrypt. Internally, the `EngineSigner` calls [`AccountProvider::sign`](https://github.com/poanetwork/parity-ethereum/blob/c159028507b8a35e1e39d6c8a11f4c72bee0a27e/ethcore/src/account_provider/mod.rs#L488) with the signer's address, which retrieves the account secret from `SecretStore` and signs the message:
```rust
let account = self.sstore.account_ref(&address)?;
// ...
Ok(self.sstore.sign_with_secret(&secret, &message)?)
```
[`sign_with_secret`](https://github.com/poanetwork/parity-ethereum/blob/c159028507b8a35e1e39d6c8a11f4c72bee0a27e/ethstore/src/secret_store.rs#L113) internally uses `ethkey` directly.

So to _decrypt_ the randomness secret, we either need to get a hold of the account secret directly (probably a bad idea), or expose additional methods for decryption from `EngineSigner` and/or `AccountProvider`, that call the corresponding secret store decryption method. (That one already exists!)

For _encryption_, we only need the _public_ key of the account. We would probably also add an `encrypt` method to  `AccountProvider`. (You'd think you wouldn't need an account provider for that, but the address itself is only the rightmost 20 bytes of the public key, so it can't be computed from the address itself. That's probably why there is an `AccountProvider::public` method at all.)

**Comment by DemiMarie** [comment]
@afck That looks good.  The implementation of ECIES looks reasonable, though I did not audit it in any detail.  My understanding (though I could be wrong) is that the same key can safely be used for both signing and encrypting.  The use of AES is a potential side-channel vulnerability, but only if a naive implementation is used or the processor does not support hardware acceleration.  Furthermore, in this case, we will only be decrypting a message we ourselves sent.  Therefore, we can make an incompatible change to the encryption format, so long as we can still recognize and decrypt old cyphertexts.

I am worried that this might break hardware wallets, which generally only provide oracle access to the signing key.

**Comment by varasev** [comment]
> For encryption, we only need the public key of the account.

@afck Are you sure about the public key? Maybe private key?

**Comment by DemiMarie** [comment]
@varasev This is asymmetric encryption, so we only need the public key.

**Comment by afck** [comment]
> I am worried that this might break hardware wallets, which generally only provide oracle access to the signing key.

You're right, that might be the `None` branch in [`sign`](https://github.com/poanetwork/parity-ethereum/blob/c159028507b8a35e1e39d6c8a11f4c72bee0a27e/ethcore/src/account_provider/mod.rs#L488). But secret store also exposes oracle encryption and decryption, I think.

**Comment by varasev** [comment]
> This is asymmetric encryption, so we only need the public key.

If the public key cannot be retrieved from validator's address, then ok.

**Comment by DemiMarie** [comment]
@varasev even if the public key could be retrieved from the validator’s address, we would be fine.  Decrypting messages requires the secret key, and we know the keccak256 hash of the secret we encrypted, so we cannot be tricked into revealing a bogus message.

**Comment by afck** [comment]
I assume the `auth_data` in the [`ecies`](https://github.com/poanetwork/parity-ethereum/blob/c159028507b8a35e1e39d6c8a11f4c72bee0a27e/ethkey/src/crypto.rs#L69) methods can basically be public—even the empty string?
(Or should we use something only known to ourselves here, like our signature of the round number?)

**Comment by varasev** [comment]
> Decrypting messages requires the secret key, and we know the keccak256 hash of the secret we encrypted, so we cannot be tricked into revealing a bogus message.

Okay, the main thing is that no one could decrypt the cipher except the validator.

**Comment by afck** [comment]
> Okay, the main thing is that no one could decrypt the cipher except the validator.

Exactly: For _decryption_ you need the secret key, for _encryption_ you need the public key, but not even the public key can be computed from the Ethereum address. Doesn't matter in this case anyway, since we only encrypt to ourselves.

**Comment by DemiMarie** [comment]
@afck `auth-data` can be public, yes.  That said, I looked at the code, and there might be a potential attack if its length is not known, since I do not see either its length or the length of the message being stored in the HMAC.  That said, I suggest we place the hash of the secret (the same hash that we revealed) there, assuming that we submit the encrypted data and the secret hash together.

We do need `auth_data` for decryption (otherwise, we will get an error saying that the message was potentially forged), so it should be stored along with the encrypted message.

I believe the public key *can* be computed from the Ethereum address + any signature that has been made (otherwise, `ecrecover` would not work).  Also, the public key is generally considered to be public information anyway (hence the name).

**Comment by afck** [comment]
Now that I realized that `AccountProvider` already has an [`account_public`](https://github.com/poanetwork/parity-ethereum/blob/c159028507b8a35e1e39d6c8a11f4c72bee0a27e/ethcore/src/account_provider/mod.rs#L371) and [`decrypt` method](https://github.com/poanetwork/parity-ethereum/blob/c159028507b8a35e1e39d6c8a11f4c72bee0a27e/ethcore/src/account_provider/mod.rs#L555), all that's missing is either:
* Keep a handle to `AccountProvider` in `AuthorityRound` or add a getter to `EngineSigner`. Do encryption with direct calls to `ethkey::crypt::ecies::encrypt`.
* Add `encrypt`, `account_public` and `decrypt` methods to `EngineSigner` (which would be a misnomer then).
* Add `encrypt_to_self` and `decrypt` methods to `EngineSigner` (also a misnomer).

**Comment by afck** [comment]
My work in progress is in https://github.com/poanetwork/parity-ethereum/tree/afck-rand-crypt. It still needs a lot of cleanups (error handling, and probably some other reorganizations), but it compiles and I think it should do the right thing. (Please double-check and test!)
I unassigned myself for now. I'll pick it up again on Monday if it's not done by then.

**Comment by DemiMarie** [comment]
Fixed by #74.
